<!DOCTYPE html>

<html lang="en" data-content_root="../">
  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" /><meta name="viewport" content="width=device-width, initial-scale=1" />

    <title>Background &#8212; Ecoevolity 0.1.0 documentation</title>
    <link rel="stylesheet" type="text/css" href="../_static/pygments.css?v=03e43079" />
    <link rel="stylesheet" type="text/css" href="../_static/basic.css?v=b08954a9" />
    <link rel="stylesheet" type="text/css" href="../_static/alabaster.css?v=f237fd5d" />
    <script src="../_static/documentation_options.js?v=01f34227"></script>
    <script src="../_static/doctools.js?v=9bcbadda"></script>
    <script src="../_static/sphinx_highlight.js?v=dc90522c"></script>
    <link rel="index" title="Index" href="../genindex.html" />
    <link rel="search" title="Search" href="../search.html" />
    <link rel="next" title="The Data" href="biallelic-character-data.html" />
    <link rel="prev" title="" href="index.html" />
   
  <link rel="stylesheet" href="../_static/custom.css" type="text/css" />
  

  
  

  </head><body>
  <div class="document">
    
      <div class="sphinxsidebar" role="navigation" aria-label="Main">
        <div class="sphinxsidebarwrapper">
<p class="logo">
  <a href="../index.html">
    <img class="logo" src="../_static/site-logo.svg" alt="Logo" />
    
  </a>
</p>






<p>
<iframe src="https://ghbtns.com/github-btn.html?user=phyletica&repo=ecoevolity&type=watch&count=true&size=large&v=2"
  allowtransparency="true" frameborder="0" scrolling="0" width="200px" height="35px"></iframe>
</p>





<h3>Navigation</h3>
<ul class="current">
<li class="toctree-l1"><a class="reference internal" href="../installation.html">Installation</a></li>
<li class="toctree-l1"><a class="reference internal" href="../ecoevolity/index.html">Ecoevolity</a></li>
<li class="toctree-l1 current"><a class="reference internal" href="index.html">Phycoeval</a><ul class="current">
<li class="toctree-l2 current"><a class="current reference internal" href="#">Background</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#the-short-story-tl-dr">The short story (tl;dr)</a></li>
<li class="toctree-l3"><a class="reference internal" href="#the-longer-story">The longer story</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="biallelic-character-data.html">The Data</a></li>
<li class="toctree-l2"><a class="reference internal" href="yaml-config.html">The Configuration File</a></li>
<li class="toctree-l2"><a class="reference internal" href="tutorials/index.html">Tutorials</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="../workshops/index.html">Workshops</a></li>
<li class="toctree-l1"><a class="reference internal" href="../changes.html">Change History</a></li>
<li class="toctree-l1"><a class="reference internal" href="../license.html">License</a></li>
<li class="toctree-l1"><a class="reference internal" href="../citation.html">Citation</a></li>
<li class="toctree-l1"><a class="reference internal" href="../acknowledgments.html">Acknowledgments</a></li>
<li class="toctree-l1"><a class="reference internal" href="../glossary.html">Glossary</a></li>
<li class="toctree-l1"><a class="reference internal" href="../zbib/references.html">Literature Cited</a></li>
</ul>

<div class="relations">
<h3>Related Topics</h3>
<ul>
  <li><a href="../index.html">Documentation overview</a><ul>
  <li><a href="index.html"></a><ul>
      <li>Previous: <a href="index.html" title="previous chapter"></a></li>
      <li>Next: <a href="biallelic-character-data.html" title="next chapter">The Data</a></li>
  </ul></li>
  </ul></li>
</ul>
</div>
<search id="searchbox" style="display: none" role="search">
  <h3 id="searchlabel">Quick search</h3>
    <div class="searchformwrapper">
    <form class="search" action="../search.html" method="get">
      <input type="text" name="q" aria-labelledby="searchlabel" autocomplete="off" autocorrect="off" autocapitalize="off" spellcheck="false"/>
      <input type="submit" value="Go" />
    </form>
    </div>
</search>
<script>document.getElementById('searchbox').style.display = "block"</script>








        </div>
      </div>
      <div class="documentwrapper">
        <div class="bodywrapper">
          

          <div class="body" role="main">
            
  <p><a class="reference external" href="index.html"><img alt="Phycoeval" src="../_images/phycoeval-logo-long.svg" style="width: 95%;" /></a></p>
<section id="background">
<span id="phycobackground"></span><h1>Background<a class="headerlink" href="#background" title="Link to this heading">¶</a></h1>
<section id="the-short-story-tl-dr">
<h2>The short story (tl;dr)<a class="headerlink" href="#the-short-story-tl-dr" title="Link to this heading">¶</a></h2>
<p>Many processes of diversification can cause
simultaneous (and potentially multifurcating) divergences,
but current phylogenetic methods for inferring rooted trees
assume evolutionary lineages diverge independently (and only bifurcate).
This leaves us without a good way to infer the patterns of shared
divergences predicted by these processes.
To solve this problem, we generalized the space of topologies considered during
phylogenetic inference to include trees with shared or multifurcating
divergences.
This allows us to jointly infer relationships, divergence times, shared
divergences, and multifurcating divergences, and test for patterns of
divergences predicted by processes of diversification that simultaneously
affect multiple lineages.</p>
</section>
<section id="the-longer-story">
<h2>The longer story<a class="headerlink" href="#the-longer-story" title="Link to this heading">¶</a></h2>
<p>Many processes of diversification can simultaneously affect multiple
lineages.
For example, below is an animation of three species of lizards co-occurring on
an island that is fragmented twice by rising sea levels.</p>
<figure class="align-center" id="id7">
<span id="shared-divs-bifurcating-gif"></span><a class="reference internal image-reference" href="../_images/slides-bifurcating.gif"><img alt="Shared divergences" src="../_images/slides-bifurcating.gif" style="width: 99%;" />
</a>
<figcaption>
<p><span class="caption-text">Island fragmentation causing shared divergences.</span><a class="headerlink" href="#id7" title="Link to this image">¶</a></p>
</figcaption>
</figure>
<p>This creates two bouts of shared divergences across the tree, indicated
by the dashed lines above.
In addition to biogeography, there are many other examples of processes
of diversification that generate patterns of shared divergences.
Instead of lizards on islands, let’s imagine three members of a gene family
residing along a region of a chromosome that gets duplicated.
This would create shared divergences across the phylogenetic history of the
gene family.
In epidemiology, when multiple infected individuals spread the pathogen to
others at a social gathering, this will create shared divergences in the
“transmission tree” of the pathogen.</p>
<p>If rising sea levels fragments the island into more than two
island, like in the animation below,
this will not only cause shared divergences among lineages,
but also multifurcations
(a lineage diverging into three or more descendants).</p>
<figure class="align-center" id="id8">
<span id="shared-divs-multifurcating-gif"></span><a class="reference internal image-reference" href="../_images/slides-multifurcating.gif"><img alt="Shared multifurcating divergences" src="../_images/slides-multifurcating.gif" style="width: 99%;" />
</a>
<figcaption>
<p><span class="caption-text">Island fragmentation causing multifurcating, shared divergences.</span><a class="headerlink" href="#id8" title="Link to this image">¶</a></p>
</figcaption>
</figure>
<p>Similarly, when an infected individual spreads a pathogen to two or
more others at a social gathering, this will create a multifurcating
divergence in the transmission tree.</p>
<p>Current phylogenetic methods for inferring rooted trees assume all divergences
are independent and bifurcating.
In other words, if we have <img class="math" src="../_images/math/2d1558f51212a7b4b1610015f8d187b354729e6c.svg" alt="\nTips"/> tips, current methods only consider
trees with <img class="math" src="../_images/math/e141357543991b67295e8b6efbcde991dcb37622.svg" alt="\nTips - 1"/> independent, bifurcating divergences.
When shared and/or multifurcating divergences were common in the system we
want to study, such tree models are over-parameterized as
<a class="reference internal" href="#true-v-current-tree-model"><span class="std std-ref">illustrated in the figure below</span></a>.
More importantly, <strong>by assuming all divergences are independent and
bifurcating, current phylogenetic methods do not allow us to test for patterns
of shared or multifurcating divergences predicted by processes of
diversification that are of interest across the life sciences</strong>.</p>
<figure class="align-center" id="id9">
<span id="true-v-current-tree-model"></span><a class="reference internal image-reference" href="../_images/gecko-trees-flipped-cropped.svg"><img alt="True versus current tree model" src="../_images/gecko-trees-flipped-cropped.svg" style="width: 99%;" />
</a>
<figcaption>
<p><span class="caption-text">When shared or multifurcating divergences have occurred, current
phylogenetic models are over-parameterized.</span><a class="headerlink" href="#id9" title="Link to this image">¶</a></p>
</figcaption>
</figure>
<p>To relax the assumption of independent, bifurcating divergences,
phycoeval
uses
a Bayesian approach to generalizing the space of tree models to allow for
shared and multifurcating divergences
<span id="id1">[<a class="reference internal" href="../zbib/references.html#id22" title="Jamie R. Oaks, Perry L. Wood, Jr., Cameron D. Siler, and Rafe M. Brown. Generalizing Bayesian phylogenetics to infer shared evolutionary events. Proceedings of the National Academy of Sciences, 119(29):e2121036119, 07 2022. URL: https://www.pnas.org/doi/abs/10.1073/pnas.2121036119, arXiv:https://www.pnas.org/doi/pdf/10.1073/pnas.2121036119, doi:10.1073/pnas.2121036119.">20</a>]</span>.
Under the generalized tree model implemented in phycoeval,
trees with <img class="math" src="../_images/math/e141357543991b67295e8b6efbcde991dcb37622.svg" alt="\nTips - 1"/> bifurcating divergences
are
<a class="reference internal" href="#tree-model-space"><span class="std std-ref">only one class of tree models in a greater space of trees</span></a>
with anywhere from <img class="math" src="../_images/math/e141357543991b67295e8b6efbcde991dcb37622.svg" alt="\nTips - 1"/> potentially shared
or multifurcating divergences.</p>
<figure class="align-center" id="id10">
<span id="tree-model-space"></span><a class="reference internal image-reference" href="../_images/four-leaf-labeled-trees-boxed-highlight-shared-grid-cropped.png"><img alt="Tree model space" src="../_images/four-leaf-labeled-trees-boxed-highlight-shared-grid-cropped.png" style="width: 99%;" />
</a>
<figcaption>
<p><span class="caption-text">The rooted topologies considered by current phylogenetic methods (within
box) and the additional tree models considered by phycoeval’s generalized
tree model.
The tree models with nonindependent (shared) divergences are highlighted in
orange.</span><a class="headerlink" href="#id10" title="Link to this image">¶</a></p>
</figcaption>
</figure>
<p>Phycoeval uses reversible-jump Markov chain Monte Carlo algorithms to sample this
generalized space of trees.
This allows joint inference of relationships, shared and multifurcating
divergences, and divergence times.</p>
<p>In phycoeval,
we coupled the generalized tree model with the “SNAPP likelihood” for directly
calculating the probability of biallelic characters given a population (or
species) phylogeny, while analytically integrating over all possible gene trees
under a coalescent model and all possible mutational histories along those gene
trees under a finite-sites model of character evolution
<span id="id2">[<a class="reference internal" href="../zbib/references.html#id558" title="David Bryant, Remco Bouckaert, Joseph Felsenstein, Noah A. Rosenberg, and Arindam Roychoudhury. Inferring species trees directly from biallelic genetic markers: bypassing gene trees in a full coalescent analysis. Molecular Biology and Evolution, 29(8):1917–1932, 2012. doi:10.1093/molbev/mss086.">3</a>, <a class="reference internal" href="../zbib/references.html#id49" title="Jamie R. Oaks. Full Bayesian comparative phylogeography from genomic data. Systematic Biology, 68(3):371–395, 05 2019. URL: https://doi.org/10.1093/sysbio/syy063, doi:10.1093/sysbio/syy063.">14</a>]</span>.
This allows us to jointly infer a species tree and shared divergences from
genomic data
<span id="id3">[<a class="reference internal" href="../zbib/references.html#id22" title="Jamie R. Oaks, Perry L. Wood, Jr., Cameron D. Siler, and Rafe M. Brown. Generalizing Bayesian phylogenetics to infer shared evolutionary events. Proceedings of the National Academy of Sciences, 119(29):e2121036119, 07 2022. URL: https://www.pnas.org/doi/abs/10.1073/pnas.2121036119, arXiv:https://www.pnas.org/doi/pdf/10.1073/pnas.2121036119, doi:10.1073/pnas.2121036119.">20</a>]</span>.</p>
</section>
</section>


          </div>
          
        </div>
      </div>
    <div class="clearer"></div>
  </div>
<div class="footer" role="contentinfo" align="center">
    <a href="https://github.com/phyletica/ecoevolity">GitHub</a> &nbsp; | &nbsp; <a href="../license.html">License</a> &nbsp; | &nbsp; <a href="../citation.html">Citation</a> &nbsp; | &nbsp; <a href="https://groups.google.com/forum/#!forum/ecoevolity-users">Users Forum</a>
</div>

  </body>
</html>